README.md
# Realtime Path Tracer for Bevy

[![crates.io](https://img.shields.io/crates/v/bevy-hikari)](https://crates.io/crates/bevy-hikari)
[![docs.rs](https://docs.rs/bevy-hikari/badge.svg)](https://docs.rs/bevy-hikari)

`bevy-hikari` is an implementation of global illumination for [Bevy](https://bevyengine.org/).

After Bevy releasing 0.8, the plugin moves to deferred hybrid path tracing.
For the old version (0.1.x) which uses voxel cone tracing with anisotropic mip-mapping, please check the `bevy-0.6` branch.

:construction: _This project is under heavily development. Certain passes are yet optimized, public interfaces may change between major versions, and minor versions may release very often._

:warning: **Notes:**
- Please run examples with `--release` flag to avoid the texture non-uniform indexing error
- Please compile this libaray in **release mode** when using as a dependency
- Supported meshes must have these 3 vertex attributes: position, normal and uv

## Bevy Version Support
| `bevy` | `bevy-hikari` |
| ------ | ------------- |
| 0.6    | 0.1           |
| 0.8    | 0.2           |
| 0.9    | 0.3           |

## Progress
- [x] Extraction and preparation of mesh assets and instances
- [ ] Asynchronous building of acceleration structures
- [x] G-Buffer generation
- [x] N-bounce indirect lighting
- [ ] Transparency
- [x] Next event estimation
- [x] Better light sampling (L-BVH + Alias table)
- [x] ReSTIR: Temporal sample reuse
- [x] ReSTIR: Spatial sample reuse
- [x] Spatiotemporal filtering
- [x] Temporal anti-aliasing
- [x] Spatial up-scaling (FSR 1.0)
- [x] Temporal up-scaling (SMAA TU4X)
- [ ] Skinned animation
- [x] HDR output
- [x] Bloom
- [ ] Hardware ray tracing (upstream related)

## Basic Usage
1. Add `HikariPlugin` to your `App` after `PbrPlugin`
2. Setup the scene with a directional light
3. Insert [`HikariSettings`](https://docs.rs/bevy-hikari/latest/bevy_hikari/struct.HikariSettings.html) component to the camera

One can configure the renderer through `HikariSettings` component on the camera entity.
Available options are:
```rust
pub struct HikariSettings {
    /// The interval of frames between sample validation passes.
    pub direct_validate_interval: usize,
    /// The interval of frames between sample validation passes.
    pub emissive_validate_interval: usize,
    /// Temporal reservoir sample count is capped by this value.
    pub max_temporal_reuse_count: usize,
    /// Spatial reservoir sample count is capped by this value.
    pub max_spatial_reuse_count: usize,
    /// Max lifetime of a reservoir sample before being replaced with new one.
    pub max_reservoir_lifetime: f32,
    /// Half angle of the solar cone apex in radians.
    pub solar_angle: f32,
    /// Count of indirect bounces.
    pub indirect_bounces: usize,
    /// Threshold for the indirect luminance to reduce fireflies.
    pub max_indirect_luminance: f32,
    /// Clear color override.
    pub clear_color: Color,
    /// Whether to do temporal sample reuse in ReSTIR.
    pub temporal_reuse: bool,
    /// Whether to do spatial sample reuse for emissive lighting in ReSTIR.
    pub emissive_spatial_reuse: bool,
    /// Whether to do spatial sample reuse for indirect lighting in ReSTIR.
    pub indirect_spatial_reuse: bool,
    /// Whether to do noise filtering.
    pub denoise: bool,
    /// Which temporal filtering implementation to use.
    pub taa: Taa,
    /// Which upscaling implementation to use.
    pub upscale: Upscale,
}
```

If you are unsure about some terms in these options, you could check [render passes](#render-passes) for details.

```rust
use bevy::prelude::*;
use bevy_hikari::prelude::*;
use std::f32::consts::PI;

fn main() {
    App::new()
        .add_plugins(DefaultPlugins)
        .add_plugin(HikariPlugin {
            remove_main_pass: true,
        })
        .add_startup_system(setup)
        .run();
}

fn setup(
    mut commands: Commands,
    mut meshes: ResMut<Assets<Mesh>>,
    mut materials: ResMut<Assets<StandardMaterial>>,
    _asset_server: Res<AssetServer>,
) {
    // Plane
    commands.spawn(PbrBundle {
        mesh: meshes.add(Mesh::from(shape::Plane { size: 5.0 })),
        material: materials.add(Color::rgb(0.3, 0.5, 0.3).into()),
        ..default()
    });
    // Cube
    commands.spawn(PbrBundle {
        mesh: meshes.add(Mesh::from(shape::Cube { size: 1.0 })),
        material: materials.add(Color::rgb(0.8, 0.7, 0.6).into()),
        transform: Transform::from_xyz(0.0, 0.5, 0.0),
        ..default()
    });

    // Only directional light is supported
    commands.spawn(DirectionalLightBundle {
        directional_light: DirectionalLight {
            illuminance: 10000.0,
            ..Default::default()
        },
        transform: Transform {
            translation: Vec3::new(0.0, 5.0, 0.0),
            rotation: Quat::from_euler(EulerRot::XYZ, -PI / 4.0, PI / 4.0, 0.0),
            ..Default::default()
        },
        ..Default::default()
    });

    // Camera
    commands.spawn((
        Camera3dBundle {
            // Set `camera_render_graph`
            camera_render_graph: CameraRenderGraph::new(bevy_hikari::graph::NAME),
            transform: Transform::from_xyz(-2.0, 2.5, 5.0).looking_at(Vec3::ZERO, Vec3::Y),
            ..Default::default()
        },
        // Add [`HikariSettings`] component to enable GI rendering
        HikariSettings::default(),
    ));
}
```

## Custom Materials
`bevy-hikari` is built based on `bevy`'s PBR standard material.
A material that is eligible for lighting needs to have surface properties a `StandardMaterial` has (base color, metallic, etc).
In order to render with custom materials,
1. Implement `Into<StandardMaterial>` for the material `CustomMaterial`, and
2. Add `GenericMaterialPlugin::<CustomMaterial>::default()` to the app
3. Add `GenericInstancePlugin::<CustomMaterial>::default()` to the app

## Screenshots
Here are the screenshots of examples.

### Simple
<img src="assets/screenshots/simple-1.png" />
<img src="assets/screenshots/simple-2.png" />

### Cornell (2 Indirect Bounces)
<img src="assets/screenshots/cornell.png">

### City (Song Dynasty)
<img src="assets/screenshots/city.png">

### Scene (WW2)
<img src="assets/screenshots/scene-1.png">
<img src="assets/screenshots/scene-2.png">
<img src="assets/screenshots/scene-3.png">
<img src="assets/screenshots/dissection/render.png">

## Render Passes
<img src="assets/screenshots/dissection/render-graph.png">

## License
Just like Bevy, all code in this repository is dual-licensed under either:

* MIT License ([LICENSE-MIT](docs/LICENSE-MIT) or [http://opensource.org/licenses/MIT](http://opensource.org/licenses/MIT))
* Apache License, Version 2.0 ([LICENSE-APACHE](docs/LICENSE-APACHE) or [http://www.apache.org/licenses/LICENSE-2.0](http://www.apache.org/licenses/LICENSE-2.0))

at your option.

## Credits
- "Fire Extinguisher" model and textures by Cameron 'cron' Fraser.
- "WW2 City Scene" from [sketchfab](https://sketchfab.com/3d-models/ww2-cityscene-carentan-inspired-639dc3d330a940a2b9d7f40542eabdf3).
- "Song Dynasty Architectures" by 唐晨.

## References
- ReSTIR: https://cs.dartmouth.edu/wjarosz/publications/bitterli20spatiotemporal.pdf
- ReSTIR GI: https://d1qx31qr3h6wln.cloudfront.net/publications/ReSTIR%20GI.pdf
- GRIS: https://d1qx31qr3h6wln.cloudfront.net/publications/sig22_GRIS.pdf
- SVGF: https://cg.ivd.kit.edu/publications/2017/svgf/svgf_preprint.pdf
- Filmic SMAA TU: https://www.activision.com/cdn/research/Dynamic_Temporal_Antialiasing_and_Upsampling_in_Call_of_Duty_v4.pdf


CHANGELOG.md
# Changelog

## [0.3.16] - 2023-2-8
### Changed
- Remove `Upscale::None` variant.

## [0.3.15] - 2022-12-24
### Changed
- Make enabling/disabling emissive spatial reuse a separate config item (default to false).

### Fixed
- Fix TAA leaving trials on the background.

## [0.3.14] - 2022-12-11
### Changed
- Make SMAA Tu4x ratio 2.0 the default setting.

## [0.3.10] - 2022-12-8
### Changed
- SMAA now re-projects and clamps previous frame pixels in a much more cleaver way,
  which greatly reduces ghosting artefacts.

### Fixed
- Fix nearest velocity implementation.

## [0.3.9] - 2022-12-7
### Changed
- SMAA now rejects samples re-projected from the previous frame also by matching sub-pixel velocity with jittering.
  This allows less aliasing for upscaled renderings in motion.
- Use nearest velocity in both SMAA and TAA.

## [0.3.8] - 2022-12-6
### Added
- Add type `HikariUniversalSettings` for disabling building of acceleration structures. ([#35](https://github.com/cryscan/bevy-hikari/issues/35))

### Fixed
- Fix more TAA ghosting for multiple upscaling ratios.

## [0.3.7] - 2022-12-5
This is mainly a bug-fix and performance upgrade.

### Changed
- Optimize data structures and bandwidth usage.

### Fixed
- Fix SMAA previous color re-projection to get less aliasing when moving.
- Fix TAA when the upscaling ratio is 1.0.

## [0.3.5] - 2022-12-4
### Changed
- Do TAA color clipping on instance miss. ([b736703](https://github.com/cryscan/bevy-hikari/tree/b73670378dab3103e4e5603e5d7e60af76188906))

### Fixed
- Suppress bloom fireflies. ([252ee68](https://github.com/cryscan/bevy-hikari/tree/252ee68964e97574c6c43f5ca7b215697e515546))

## [0.3.4] - 2022-12-2
### Changed
- Apply firefly filtering to emissive lighting. ([e7c8c52](https://github.com/cryscan/bevy-hikari/tree/e7c8c52526404724d8a2d6d55e22bc62f0e9ae02))
- Invalidate previous spatial reservoir if the corresponding temporal reservoir is invalidated to speedup light updating.

## [0.3.3] - 2022-12-1
### Changed
- Improve SMAA Tu4x quality when moving by sampling previous full scale rendering. ([065ca1c](https://github.com/cryscan/bevy-hikari/tree/065ca1ce5d10c3dca69f415e7a9c46072160a68a))
- Unroll loops in denoiser. ([4ecfbf5](https://github.com/cryscan/bevy-hikari/tree/4ecfbf54425142ac934df7dcb9759209f95c4e6e))

### Fixed
- Spatial reusing kernel jitter each frame to avoid non-convergence. ([aca0001](https://github.com/cryscan/bevy-hikari/tree/aca00016e9b99d3582b73ea51b93cb54fdf50779))

## [0.3.2] - 2022-11-29
### Changed
- **Breaking:** Add `hikari` render graph back.
  Users need to set `camera_render_graph` in their `Camera3dBundle` to `bevy_hikari::graph::NAME` in order to activate path tracing.
- Bump `bevy` from v0.9.0 to v0.9.1. ([863c73f](https://github.com/cryscan/bevy-hikari/tree/863c73fe5f649dc2a670eb6cae6817e02c6a1973))
- Change how TAA deals with dis-occlusion from color clipping to exponential blending.
  This introduces some ghosting but yields better upscaled results.
- Render prepass textures (G-Buffers) in upscaled resolution. ([bb37f6a](https://github.com/cryscan/bevy-hikari/tree/bb37f6a7d085edd475d2142bfcdc5a3176ee3e10))

### Added
- Construct and sample light BVH when rendering emissive lights. ([#86](https://github.com/cryscan/bevy-hikari/pull/86))
- Firefly filtering for indirect denoiser. ([8e87987](https://github.com/cryscan/bevy-hikari/tree/8e8798768f082233d8b8c39fcabff4a47fccb38e))

### Removed
- Disable denoiser's temporal accumulation. ([97c4081](https://github.com/cryscan/bevy-hikari/tree/97c4081df6dee24d6e11df2ea0059a4126795d62))

[0.3.16]: https://github.com/cryscan/bevy-hikari/commits/v0.3.16
[0.3.15]: https://github.com/cryscan/bevy-hikari/commits/v0.3.15
[0.3.14]: https://github.com/cryscan/bevy-hikari/commits/v0.3.14
[0.3.10]: https://github.com/cryscan/bevy-hikari/commits/v0.3.10
[0.3.9]: https://github.com/cryscan/bevy-hikari/commits/v0.3.9
[0.3.8]: https://github.com/cryscan/bevy-hikari/commits/v0.3.8
[0.3.7]: https://github.com/cryscan/bevy-hikari/commits/v0.3.7
[0.3.5]: https://github.com/cryscan/bevy-hikari/commits/v0.3.5
[0.3.4]: https://github.com/cryscan/bevy-hikari/commits/v0.3.4
[0.3.3]: https://github.com/cryscan/bevy-hikari/commits/v0.3.3
[0.3.2]: https://github.com/cryscan/bevy-hikari/commits/v0.3.2

src/main.rs
use crate::voxel_cone_tracing::Volume;
use bevy::{
    diagnostic::{FrameTimeDiagnosticsPlugin, LogDiagnosticsPlugin},
    prelude::*,
};
use bevy_flycam::{FlyCam, NoCameraPlayerPlugin};
use std::f32::consts::PI;

mod voxel_cone_tracing;

fn main() {
    let mut app = App::new();

    app.insert_resource(Msaa { samples: 4 })
        .add_plugins(DefaultPlugins)
        .add_plugin(NoCameraPlayerPlugin)
        .add_plugin(FrameTimeDiagnosticsPlugin::default())
        .add_plugin(LogDiagnosticsPlugin::default())
        .add_plugin(voxel_cone_tracing::VoxelConeTracingPlugin)
        .add_startup_system(setup)
        .add_system(controller_system)
        .add_system(light_rotate_system);

    // bevy_mod_debugdump::print_render_graph(&mut app);

    app.run();
}

#[derive(Component)]
struct Controller;

#[derive(Component)]
struct DirectionalLightTarget;

/// Set up a simple 3D scene
fn setup(
    mut commands: Commands,
    asset_server: Res<AssetServer>,
    mut meshes: ResMut<Assets<Mesh>>,
    mut materials: ResMut<Assets<StandardMaterial>>,
) {
    // Ground
    commands.spawn_bundle(PbrBundle {
        mesh: meshes.add(Mesh::from(shape::Cube::default())),
        material: materials.add(StandardMaterial {
            base_color: Color::rgb(0.6, 1.0, 0.6),
            perceptual_roughness: 1.0,
            ..Default::default()
        }),
        transform: Transform {
            translation: Vec3::new(0.0, -2.0, 0.0),
            rotation: Quat::IDENTITY,
            scale: Vec3::new(5.0, 0.5, 5.0),
        },
        ..Default::default()
    });
    // Right
    commands.spawn_bundle(PbrBundle {
        mesh: meshes.add(Mesh::from(shape::Cube::default())),
        material: materials.add(StandardMaterial {
            base_color: Color::rgb(1.0, 0.5, 0.5),
            perceptual_roughness: 1.0,
            ..Default::default()
        }),
        transform: Transform {
            translation: Vec3::new(2.0, 0.0, 0.0),
            rotation: Quat::from_rotation_z(PI / 2.0),
            scale: Vec3::new(5.0, 0.5, 4.0),
        },
        ..Default::default()
    });
    // Left
    commands.spawn_bundle(PbrBundle {
        mesh: meshes.add(Mesh::from(shape::Cube::default())),
        material: materials.add(StandardMaterial {
            base_color: Color::rgb(1.0, 1.0, 1.0),
            perceptual_roughness: 1.0,
            ..Default::default()
        }),
        transform: Transform {
            translation: Vec3::new(-2.0, 0.0, 0.0),
            rotation: Quat::from_rotation_z(PI / 2.0),
            scale: Vec3::new(5.0, 0.5, 4.0),
        },
        ..Default::default()
    });
    // Back
    commands.spawn_bundle(PbrBundle {
        mesh: meshes.add(Mesh::from(shape::Cube::default())),
        material: materials.add(StandardMaterial {
            base_color: Color::rgb(0.2, 0.8, 1.0),
            perceptual_roughness: 1.0,
            ..Default::default()
        }),
        transform: Transform {
            translation: Vec3::new(0.0, 0.0, -2.0),
            rotation: Quat::from_rotation_x(PI / 2.0),
            scale: Vec3::new(4.0, 0.5, 5.0),
        },
        ..Default::default()
    });
    // Top
    commands.spawn_bundle(PbrBundle {
        mesh: meshes.add(Mesh::from(shape::Cube::default())),
        material: materials.add(StandardMaterial {
            base_color: Color::rgb(0.9, 0.9, 0.7),
            perceptual_roughness: 1.0,
            ..Default::default()
        }),
        transform: Transform {
            translation: Vec3::new(1.0, 2.0, -0.0),
            rotation: Quat::IDENTITY,
            scale: Vec3::new(1.0, 0.5, 4.0),
        },
        ..Default::default()
    });
    commands.spawn_bundle(PbrBundle {
        mesh: meshes.add(Mesh::from(shape::Cube::default())),
        material: materials.add(StandardMaterial {
            base_color: Color::rgb(0.9, 0.9, 0.7),
            perceptual_roughness: 1.0,
            ..Default::default()
        }),
        transform: Transform {
            translation: Vec3::new(-1.0, 2.0, -0.0),
            rotation: Quat::IDENTITY,
            scale: Vec3::new(1.0, 0.5, 4.0),
        },
        ..Default::default()
    });

    // Target
    commands
        .spawn_bundle(PbrBundle {
            mesh: meshes.add(Mesh::from(shape::Icosphere {
                radius: 0.1,
                ..Default::default()
            })),
            material: materials.add(StandardMaterial {
                base_color: Color::rgb(1.0, 1.0, 1.0),
                perceptual_roughness: 1.0,
                emissive: Color::rgb(1.0, 1.0, 1.0),
                ..Default::default()
            }),
            transform: Transform::from_xyz(2.0, 0.0, -2.0),
            ..Default::default()
        })
        .insert(DirectionalLightTarget);

    const HALF_SIZE: f32 = 5.0;
    commands.spawn_bundle(DirectionalLightBundle {
        directional_light: DirectionalLight {
            illuminance: 10000.0,
            shadow_projection: OrthographicProjection {
                left: -HALF_SIZE,
                right: HALF_SIZE,
                bottom: -HALF_SIZE,
                top: HALF_SIZE,
                near: -10.0 * HALF_SIZE,
                far: 10.0 * HALF_SIZE,
                ..Default::default()
            },
            shadows_enabled: true,
            ..Default::default()
        },
        transform: Transform {
            translation: Vec3::new(0.0, 5.0, 0.0),
            rotation: Quat::from_euler(EulerRot::XYZ, -PI / 8.0, -PI / 4.0, 0.0),
            ..Default::default()
        },
        ..Default::default()
    });

    commands
        .spawn()
        .insert_bundle((
            Transform {
                translation: Vec3::new(0.0, -1.0, 0.0),
                rotation: Quat::IDENTITY,
                scale: Vec3::new(2.0, 2.0, 2.0),
            },
            GlobalTransform::default(),
            Controller,
        ))
        .with_children(|parent| {
            parent.spawn_scene(asset_server.load("models/FlightHelmet/FlightHelmet.gltf#Scene0"));
        });

    // Camera
    commands
        .spawn_bundle(PerspectiveCameraBundle {
            transform: Transform::from_xyz(0.0, 0.0, 5.0).looking_at(Vec3::ZERO, Vec3::Y),
            ..Default::default()
        })
        .insert(Volume::new(
            Vec3::new(-2.5, -2.5, -2.5),
            Vec3::new(2.5, 2.5, 2.5),
        ))
        .insert(FlyCam);

    commands.spawn_bundle(UiCameraBundle::default());
}

fn controller_system(
    windows: Res<Windows>,
    keyboard_input: Res<Input<KeyCode>>,
    time: Res<Time>,
    mut controller_query: Query<&mut Transform, With<Controller>>,
) {
    let right = Vec3::X;
    let forward = -Vec3::Z;
    let speed = 2.0;

    let window = windows.get_primary().unwrap();

    for mut transform in controller_query.iter_mut() {
        if !window.cursor_locked() {
            if keyboard_input.pressed(KeyCode::W) {
                transform.translation += forward * speed * time.delta_seconds();
            }
            if keyboard_input.pressed(KeyCode::A) {
                transform.translation -= right * speed * time.delta_seconds();
            }
            if keyboard_input.pressed(KeyCode::S) {
                transform.translation -= forward * speed * time.delta_seconds();
            }
            if keyboard_input.pressed(KeyCode::D) {
                transform.translation += right * speed * time.delta_seconds();
            }
            if keyboard_input.pressed(KeyCode::E) {
                transform.translation += Vec3::Y * speed * time.delta_seconds();
            }
            if keyboard_input.pressed(KeyCode::Q) {
                transform.translation -= Vec3::Y * speed * time.delta_seconds();
            }
        }

        let speed = 0.7;
        transform.rotation *=
            Quat::from_euler(EulerRot::XYZ, 0.0, speed * time.delta_seconds(), 0.0);
    }
}

fn light_rotate_system(
    keyboard_input: Res<Input<KeyCode>>,
    time: Res<Time>,
    mut query: QuerySet<(
        QueryState<&mut Transform, With<DirectionalLight>>,
        QueryState<&mut Transform, With<DirectionalLightTarget>>,
    )>,
) {
    let mut target_query = query.q1();
    let mut target = target_query.single_mut();

    let right = Vec3::X;
    let forward = -Vec3::Z;
    let speed = 4.0;

    if keyboard_input.pressed(KeyCode::Up) {
        target.translation += forward * speed * time.delta_seconds();
    }
    if keyboard_input.pressed(KeyCode::Down) {
        target.translation -= forward * speed * time.delta_seconds();
    }
    if keyboard_input.pressed(KeyCode::Left) {
        target.translation -= right * speed * time.delta_seconds();
    }
    if keyboard_input.pressed(KeyCode::Right) {
        target.translation += right * speed * time.delta_seconds();
    }
    let target = target.translation;

    let mut light_query = query.q0();
    let mut light = light_query.single_mut();

    light.look_at(target, Vec3::Y);
}
